Add Expo template featuring RTK 2.0

aryaemami59 opened the pull request:
This PR:

- [X] Adds template for Expo.
- [X] Replaces examples of `createAsyncThunk` with `create.asyncThunk`
- [X] Adds examples of RTK Query using `createApi`, and auto-generated React hooks
- [X] Resolves #91
- [X] Adds examples of the new `combineSlices` API
- [X] Adds `setupListeners` API
- [X] Replaces inline selectors with selectors defined in `createSlice.selectors`
- [X] Resolves #90 by adding integration tests

What else can be done:

- I would like to at some point add an example for Reselect
- We can also add some unit tests for RTK Query

## Additional Notes:

In adapting this template for `create-expo`, several modifications were necessary. Unlike more complex scaffolding tools like `create-react-app` or `react-native init`, `create-expo` has a straightforward approach. It requires the template to be an NPM package, but with specific considerations:
  
  ### Modifications in `package.json`
  `create-expo` automatically alters the package.json file in certain ways:
  
  - Removes existing `description`, `tags`, and `repository` fields.
  - Changes `version` to `1.0.0`.
  - Sets `private` to `true`.
  - Changes `name` to name of the project.
  
For reference you can see what they do here: https://github.com/expo/expo/blob/2dc48d7130d14d475c8aa79be22bf83c2593a385/packages/create-expo/src/Template.ts#L132C3-L139C33

  ### Handling of `.gitignore`
  `create-expo`  has a specific way of processing `.gitignore` files:
  
  - It generates a default `.gitignore` if one doesn’t exist.
  - Existing `.gitignore` files are replaced with this default.
  - If a `gitignore` file is present, `create-expo` uses this instead.
  
  So basically `gitignore` becomes the new `.gitignore`. So we can use `.gitignore` to ignore lock files during development, which is why I added it.
  
  Reference: https://github.com/expo/expo/blob/b1924a0637b30e94a2b2fa82eca4d12702c3684a/packages/create-expo/src/createFileTransform.ts#L44C5-L48C6
  
This template can be tested locally by running:

```bash
npx create-expo my-app -t file:<relative or absolute path to template> --no-install
```
I suggest using `--no-install` to bypass the time-consuming Expo installations if you're primarily interested in the project's initial structure.

aryaemami59 committed:
Add Expo template

aryaemami59 committed:
Add "@typescript-eslint/no-restricted-imports" rule to `.eslintrc.json`

aryaemami59 committed:
Remove `package-lock.json` file

aryaemami59 committed:
Update dependencies

aryaemami59 committed:
Rename `createSliceWithThunks.ts` to `createAppSlice.ts` to be consistent with RTK docs

aryaemami59 committed:
Update dependencies

aryaemami59 committed:
Add `hooks.withTypes()` to `hooks.ts`

aryaemami59 committed:
Update `App.test.tsx` to use aria-labels instead of test-ids

aryaemami59 committed:
Rename `.gitignore` to `gitignore`

aryaemami59 committed:
Copy `gitignore` from `create-expo`

aryaemami59 committed:
Fix confusing format command

aryaemami59 committed:
Add `eslint-config-prettier` to dev dependencies

aryaemami59 committed:
Modify ESLint config to prevent potential conflict with prettier

markerikson committed:
Add comment to makeStore

markerikson commented:
I have no idea if this actually works :) I'll assume it does based on your work and the similarities to the other templates.

Does Expo add a default README when you use a template? If not, we should add one.

aryaemami59 commented:
@markerikson It doesn't, I'll add one.

aryaemami59 committed:
Merge branch 'master' of https://github.com/reduxjs/redux-templates into add-expo-ts-template

aryaemami59 committed:
Merge branch 'master' of https://github.com/reduxjs/redux-templates into add-expo-ts-template

aryaemami59 committed:
Add `README` file

aryaemami59 committed:
Update `package.json` for compatibility with `create-expo`

- Added `description` and `repository` fields to enhance metadata.
- Removed the `private` field to allow package publication.
- Adapted to `create-expo` behavior, which modifies `package.json` by:
  - Removing existing `description`, `tags`, and `repository` fields.
  - Standardizing `version` to `1.0.0`.
  - Setting `private` to `true`.
  - Updating `name` to match the project name.
- Changes ensure `expo-template-redux-typescript` can be published and utilized with `create-expo`.

aryaemami59 committed:
Add `.gitignore` for `create-expo` compatibility

- Added a `.gitignore` file for use during template development.
- The previously added `gitignore` file becomes the active `.gitignore` post-template usage.
- This dual-file strategy accounts for `create-expo`'s handling of `.gitignore` files:
  - Provides a default `.gitignore` if none exists.
  - Overrides an existing `.gitignore` with a default version.
  - Utilizes a `gitignore` file, if present, instead of creating a new one.
- The setup ensures our custom rules are maintained when the template is employed.
